### PR TITLE
Manual OTA update

### LINKED ISSUE #97: Implement manual OTA in the backend

### PR DESCRIPTION
**Pull request by rbino**
Allow pushing a single OTA update on a specific device, and expose the progress using an OTA operation

Close #97

### CODE REVIEW

File: backend/lib/edgehog/os_management.ex [hunk 1]
```diff
@@ -0,0 +1,198 @@
+#
+# This file is part of Edgehog.
+#
+# Copyright 2021 SECO Mind Srl
```
Comment (Pavinati) [inline backend/lib/edgehog/os_management.ex:4]: 2022

File: backend/lib/edgehog/astarte/device/ota_request/behaviour.ex [hunk 2]
```diff
@@ -0,0 +1,29 @@
+#
+# This file is part of Edgehog.
+#
+# Copyright 2021 SECO Mind Srl
```
Comment (Pavinati) [inline backend/lib/edgehog/astarte/device/ota_request/behaviour.ex:4]: 2022

File: backend/lib/edgehog/astarte/device/ota_request.ex [hunk 3]
```diff
@@ -0,0 +1,31 @@
+#
+# This file is part of Edgehog.
+#
+# Copyright 2021 SECO Mind Srl
```
Comment (Pavinati) [inline backend/lib/edgehog/astarte/device/ota_request.ex:4]: 2022

File: backend/lib/edgehog_web/schema/os_management_types.ex [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  @desc "An OTA update operation"
+  node object(:ota_operation) do
+    @desc "The URL of the base image being installed on the device"
+    field :base_image_url, non_null(:string)
+
+    @desc "The current status of the operation"
+    field :status, non_null(:ota_operation_status)
+
+    @desc "The current status code of the operation"
+    field :status_code, :string
```
Comment (Pavinati) [inline backend/lib/edgehog_web/schema/os_management_types.ex:48]: this is an enum as far as i know. Those are the values used by the ESP32 firmware
```
"OTAErrorNetwork"
"OTAErrorNvs"
"OTAAlreadyInProgress"
"OTAFailed"
"OTAErrorDeploy"
"OTAErrorBootWrongPartition"
```

File: backend/priv/repo/migrations/20211223141031_create_ota_operations.exs [hunk 5]
```diff
@@ -0,0 +1,27 @@
+defmodule Edgehog.Repo.Migrations.CreateOtaOperations do
```
[the remark this answers is not in the document]
Comment (rbino) [reply backend/priv/repo/migrations/20211223141031_create_ota_operations.exs:1]: We don't put license headers in migrations

File: backend/lib/edgehog_web/schema/astarte_types.ex [hunk 6]
```diff
@@ -322,6 +322,12 @@ defmodule EdgehogWeb.Schema.AstarteTypes do
       resolve &Resolvers.Astarte.fetch_os_info/3
       middleware Middleware.ErrorHandler
     end
+
+    @desc "The existing OTA operations for this device"
+    field :ota_operations, non_null(list_of(:ota_operation)) do
```
Comment (Pavinati) [inline backend/lib/edgehog_web/schema/astarte_types.ex:327]: `non_null(list_of(non_null(:ota_operation)))`
Members of this list shouldn't be null

**Review by bettio — APPROVED** [review]

**Review by davidebriani — APPROVED** [review]

**Review by szakhlypa — APPROVED** [review]

**Pull request merged**